File: api-quarkus/src/main/java/br/ufrn/musi/adaptadores/web/Parametros.java
package br.ufrn.musi.adaptadores.web;

import br.ufrn.musi.aplicacao.PedidoDePagina;
import br.ufrn.musi.dominio.Dominio.Ordenacao;
import jakarta.ws.rs.BadRequestException;
import java.util.Arrays;

/**
 * Leitura dos parâmetros de consulta e de caminho.
 *
 * Os parâmetros numéricos chegam como `String` de propósito: com `@QueryParam Integer`,
 * um `?pagina=abc` vira `404` (é o que a especificação Jakarta REST manda), e o certo é
 * `400`. Mesmo comportamento do Parametros.kt do lado Ktor.
 */
final class Parametros {
    private Parametros() {}

    static Integer inteiro(String nome, String valor) {
        if (valor == null || valor.isBlank()) return null;
        try {
            return Integer.valueOf(valor);
        } catch (NumberFormatException e) {
            throw new BadRequestException("`" + nome + "` deve ser um inteiro: `" + valor + "`");
        }
    }

    static long longo(String nome, String valor) {
        try {
            return Long.parseLong(valor);
        } catch (NumberFormatException e) {
            throw new BadRequestException("`" + nome + "` deve ser um inteiro: `" + valor + "`");
        }
    }

    static String texto(String valor) {
        return valor == null || valor.isBlank() ? null : valor;
    }

    static PedidoDePagina pagina(String pagina, String tamanho) {
        Integer p = inteiro("pagina", pagina);
        Integer t = inteiro("tamanho", tamanho);
        return new PedidoDePagina(p == null ? 0 : p, t == null ? PedidoDePagina.PADRAO : t);
    }

    /** `?ordem=` com os valores de `Ordenacao`, em minúsculas e com hífen (ADR-0002). */
    static Ordenacao ordenacao(String valor) {
        if (texto(valor) == null) return Ordenacao.TITULO;
        return Arrays.stream(Ordenacao.values())
            .filter(o -> paraParametro(o).equals(valor))
            .findFirst()
            .orElseThrow(() -> new BadRequestException("`ordem` deve ser um de "
                + Arrays.stream(Ordenacao.values()).map(Parametros::paraParametro).toList() + ": `" + valor + "`"));
    }

    static String paraParametro(Ordenacao o) {
        return o.name().toLowerCase().replace('_', '-');
    }
}

File: api-ktor/src/main/kotlin/br/ufrn/musi/adaptadores/web/Parametros.kt
package br.ufrn.musi.adaptadores.web

import br.ufrn.musi.aplicacao.PedidoDePagina
import br.ufrn.musi.dominio.Ordenacao
import io.ktor.server.application.ApplicationCall
import io.ktor.server.plugins.BadRequestException

/**
 * Leitura dos parâmetros de caminho e de consulta.
 *
 * Tipo errado (`?pagina=abc`, `/anotacoes/xyz`) é requisição malformada: `400`. Valor
 * bem formado que viola regra (`?tamanho=500`) é `422`, decidido pela aplicação.
 */
fun ApplicationCall.inteiro(nome: String): Int? =
    request.queryParameters[nome]?.let { it.toIntOrNull() ?: throw BadRequestException("`$nome` deve ser um inteiro: `$it`") }

fun ApplicationCall.texto(nome: String): String? = request.queryParameters[nome]?.takeIf { it.isNotBlank() }

fun ApplicationCall.caminho(nome: String): String = parameters[nome]!!

fun ApplicationCall.caminhoLong(nome: String): Long =
    parameters[nome]!!.let { it.toLongOrNull() ?: throw BadRequestException("`$nome` deve ser um inteiro: `$it`") }

fun ApplicationCall.pedidoDePagina() =
    PedidoDePagina(inteiro("pagina") ?: 0, inteiro("tamanho") ?: PedidoDePagina.PADRAO)

/**
 * `?ordem=` com os valores de `Ordenacao`, em minúsculas e com hífen. Sem padrão oculto de
 * relevância: a ordenação é sempre declarada por quem consulta (ADR-0002).
 */
fun ApplicationCall.ordenacao(): Ordenacao {
    val valor = texto("ordem") ?: return Ordenacao.TITULO
    return Ordenacao.entries.firstOrNull { it.paraParametro() == valor }
        ?: throw BadRequestException("`ordem` deve ser um de ${Ordenacao.entries.map { it.paraParametro() }}: `$valor`")
}

fun Ordenacao.paraParametro() = name.lowercase().replace('_', '-')
